demonstrations/tutorial_gradient-methods.py

```diff
@@ -0,0 +1,712 @@
+# adapted for any set of gates with generators that have two eigenvalues.
+#
+# Note that parameter-shift rules can also be derived for higher order derivatives, for instance,
+# to compute the Hessian (via :func:`~.pennylane.gradients.param_shift_hessian`) or the
+# Fisher-information matrix (via :func:`~.pennylane.metric_tensor`) of the cost function [#Mari]_.
+#
+
+@qml.qnode(dev, diff_method="parameter-shift")
+def cost(theta):
+    qfunc(theta)
+    return qml.expval(qml.PauliZ(1))
+
+grad_fn = qml.grad(cost)
+grad_fn(params)
+
+######################################################################
+# Note how this gives the exact same result as the ``backprop`` method, i.e. it is exact. On quantum
+# hardware, however, one would have to estimate the shifted cost function values with a finite number of
+# shots, which leads to statistical errors.
+#
+#
+# *PennyLane Demo:* `Quantum gradients with backpropagation <tutorial_backprop.html>`__
+#
+# *PennyLane Glossary:* `Parameter-shift Rule <../glossary/parameter_shift.html>`__
+#
+
+######################################################################
+# Hadamard Test
+# ^^^^^^^^^^^^^
+#
+# When writing out the derivative of :math:`C` explicitly as
+#
+# .. math:: \partial_j C = \langle 0 \vert \partial_j U^\dagger(\theta) \,A\, U(\theta)\vert 0 \rangle + \langle 0\vert U^\dagger(\theta) \,A\, \partial_j U(\theta)\vert 0 \rangle,
+# 
+# we can observe that it is equal to a very similar expression as :math:`C` itself. In fact, since
+# :math:`\partial_j U^\dagger = - (\partial_j U)^\dagger` (assuming :math:`\theta_j` enters the
+# circuit in the form
+# :math:`\mathrm{e}^{-\mathrm{i}\theta_j P}`, where :math:`P` is an arbitary Hermitian operator),
+# the derivative of the cost function is simply given
+# by the imaginary part of
+# 
+# .. math:: 2\, \langle 0\vert U^\dagger(\theta) \,A\, \partial_j U(\theta)\vert 0 \rangle.
+#
+# This means we need to compute the imaginary part of the overlap of two quantum states, which can
+# be done via the Hadamard test [#Guerreschi]_. Note that the Hadamard test requires an ancilla
+# qubit. For our example circuit above this means we have to use a device which has three qubits.
+#
+
+dev_3qubits = qml.device("default.qubit", wires=3)
+
+@qml.qnode(dev_3qubits, diff_method="hadamard")
+def cost(theta):
+    qfunc(theta)
+    return qml.expval(qml.PauliZ(1))
+
+grad_fn = qml.grad(cost)
+grad_fn(params)
+
+######################################################################
+# Note that this gives us again the exact gradient, just as in the example for the parameter-shift
+# rule. Now, let’s take a look at the two circuits which compute the derivative. We see that we have a controlled
+# :math:`X` and a controlled :math:`Y` gate, corresponding to the two parametrized gates in our
+# circuit.
+#
+
+(grad_tape1, grad_tape2), _ = qml.gradients.hadamard_grad(cost.tape)
+
+qml.drawer.tape_mpl(grad_tape1)
+qml.drawer.tape_mpl(grad_tape2)
```

Comment:
I'd suggest to provide the wire order to the drawing function to make the tapes more relatable to another (and to the original tape):
```suggestion
qml.drawer.tape_mpl(grad_tape1, wire_order=dev_3qubits.wires)
qml.drawer.tape_mpl(grad_tape2, wire_order=dev_3qubits.wires)
```